### PR TITLE
server: support http index mvcc interface

### PR DESCRIPTION
**Pull request by jackysp**
@AndreMouche PTAL

### CODE REVIEW

File: server/region_handler.go [hunk 1]
```diff
@@ -694,6 +740,139 @@ func (t *regionHandlerTool) getMvccByStartTs(startTS uint64, startKey, endKey []
...
+	kvResp, err := t.store.SendReq(t.bo, tikvReq, keyLocation.Region, time.Minute)
+	log.Info(string(encodeKey), keyLocation.Region, string(keyLocation.StartKey), string(keyLocation.EndKey), kvResp, err)
+
+	if err != nil {
+		return nil, err
+	}
+	return kvResp.MvccGetByKey, nil
+}
+
+func (t *regionHandlerTool) formValue2DatumRow(values url.Values, idxCols []*model.ColumnInfo) ([]types.Datum, error) {
```
Comment (shenli) [inline server/region_handler.go]: Add comment for the function.

File: server/region_handler_test.go [hunk 2]
```diff
@@ -32,6 +32,7 @@ import (
 	"github.com/pingcap/tidb/store/tikv/mocktikv"
 	"github.com/pingcap/tidb/tablecodec"
 	"github.com/pingcap/tidb/util/codec"
+	"io"
```
Comment (shenli) [inline server/region_handler_test.go]: Move this before "github.com/xxxxx".

File: server/region_handler.go [hunk 3]
```diff
@@ -592,8 +608,38 @@ func (rh mvccTxnHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	}
 }
 
+func (rh mvccTxnHandler) handleMvccGetByIdx(params map[string]string, values url.Values) (interface{}, error) {
```
Comment (shenli) [inline server/region_handler.go]: Add comment.

File: server/region_handler.go [hunk 4]
```diff
@@ -694,6 +740,139 @@ func (t *regionHandlerTool) getMvccByStartTs(startTS uint64, startKey, endKey []
...
+	for i, col := range idxCols {
+		var d types.Datum
+
+		vals, ok := values[col.Name.String()]
+		if !ok {
+			return nil, errors.BadRequestf("Missing value for index column %s.", col.Name.String())
+		}
+
+		val := vals[0]
+		if len(val) > 0 {
```
Comment (AndreMouche) [inline server/region_handler.go]: what's the difference between `NULL` and empty string?

File: server/region_handler.go [hunk 5]
```diff
@@ -694,6 +740,139 @@ func (t *regionHandlerTool) getMvccByStartTs(startTS uint64, startKey, endKey []
...
+	if err != nil {
+		return nil, err
+	}
+	return kvResp.MvccGetByKey, nil
+}
+
+func (t *regionHandlerTool) formValue2DatumRow(values url.Values, idxCols []*model.ColumnInfo) ([]types.Datum, error) {
+	data := make([]types.Datum, len(idxCols))
+	for i, col := range idxCols {
+		var d types.Datum
```
Comment (AndreMouche) [inline server/region_handler.go]: why not use data[i] directly?

File: server/region_handler.go [hunk 6]
```diff
@@ -694,6 +740,139 @@ func (t *regionHandlerTool) getMvccByStartTs(startTS uint64, startKey, endKey []
...
+	}
+	handle, err := strconv.ParseInt(handleStr, 10, 64)
+	if err != nil {
+		return nil, err
+	}
+	encodeKey, _, err := idx.GenIndexKey(idxRow, handle)
+	if err != nil {
+		return nil, err
+	}
+	keyLocation, err := t.regionCache.LocateKey(t.bo, encodeKey)
```
Comment (AndreMouche) [inline server/region_handler.go]: Repeated with L676-L693

File: server/region_handler_test.go [hunk 7]
```diff
@@ -282,6 +300,46 @@ func (ts *TidbRegionHandlerTestSuite) TestGetMvcc(c *C) {
 	err = decoder.Decode(&data2)
 	c.Assert(err, IsNil)
 	c.Assert(data2, DeepEquals, data)
+
+	// tests for normal index key
```
Comment (AndreMouche) [inline server/region_handler_test.go]: Could we define an independent function to test index key?

File: server/region_handler_test.go [hunk 8]
```diff
@@ -282,6 +300,46 @@ func (ts *TidbRegionHandlerTestSuite) TestGetMvcc(c *C) {
...
+	resp, err = http.Get("http://127.0.0.1:10090/mvcc/index/tidb/test/idx1/1?a=1&b=2")
+	c.Assert(err, IsNil)
+	decodeKeyMvcc(resp.Body, c, true)
+
+	resp, err = http.Get("http://127.0.0.1:10090/mvcc/index/tidb/test/idx2/1?a=1&b=2")
+	c.Assert(err, IsNil)
+	decodeKeyMvcc(resp.Body, c, true)
+
+	// tests for index key which includes null
```
Comment (AndreMouche) [inline server/region_handler_test.go]: Any tests for index key that contains an empty string?

File: server/region_handler.go (hunk 3 above)
Comment (jackysp) [reply server/region_handler.go]: Done.

File: server/region_handler.go (hunk 1 above)
Comment (jackysp) [reply server/region_handler.go]: Done.

File: server/region_handler_test.go (hunk 2 above)
Comment (jackysp) [reply server/region_handler_test.go]: Done.

File: server/region_handler.go (hunk 6 above)
Comment (jackysp) [reply server/region_handler.go]: Done.

File: server/region_handler.go (hunk 5 above)
Comment (jackysp) [reply server/region_handler.go]: Done.

File: server/region_handler.go (hunk 4 above)
Comment (jackysp) [reply server/region_handler.go]: Done.

File: server/region_handler_test.go (hunk 7 above)
Comment (jackysp) [reply server/region_handler_test.go]: Done.

File: server/region_handler_test.go (hunk 8 above)
Comment (jackysp) [reply server/region_handler_test.go]: Done.

**Comment by jackysp** [comment]
PTAL @shenli and @AndreMouche .

**Comment by jackysp** [comment]
@disksing

File: server/region_handler.go [hunk 9]
```diff
@@ -768,6 +874,125 @@ func (t *regionHandlerTool) getMvccByStartTs(startTS uint64, startKey, endKey []
...
+				data[i].SetMysqlBit([]byte(val))
+			case mysql.TypeJSON:
+				elem, err := tjson.ParseFromString(val)
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
+				data[i].SetMysqlJSON(elem)
+			default:
+				return nil, errors.NotSupportedf("Not supported datum type %d.", col.Tp)
+			}
```
Comment (zimulala) [inline server/region_handler.go]: Can we use the function of `ConvertTo` instead of line908-987?

File: server/region_handler.go [hunk 10]
```diff
@@ -768,6 +874,125 @@ func (t *regionHandlerTool) getMvccByStartTs(startTS uint64, startKey, endKey []
 	}
 }
 
+func (t *regionHandlerTool) getMvccByIdxValue(idx table.Index, values url.Values, idxCols []*model.ColumnInfo, handleStr string) (*kvrpcpb.MvccGetByKeyResponse, error) {
+	idxRow, err := t.formValue2DatumRow(values, idxCols)
+	if err != nil {
+		return nil, err
```
Comment (zimulala) [inline server/region_handler.go]: Do we need to add the error trace? Some places have not added in this PR.

File: server/region_handler.go [hunk 11]
```diff
@@ -647,17 +653,82 @@ func (ir RegionFrameRange) lastTableID() int64 {
...
+			continue
+		}
+		if i := strings.Index(key, "="); i >= 0 {
+			value := ""
+			var err1 error
+			key, value = key[:i], key[i+1:]
+			key, err1 = url.QueryUnescape(key)
+			if err1 != nil {
+				if err == nil {
+					err = err1
```
Comment (shenli) [inline server/region_handler.go]: Could we return err here?

File: server/region_handler.go [hunk 12]
```diff
@@ -647,17 +653,82 @@ func (ir RegionFrameRange) lastTableID() int64 {
...
+				}
+				continue
+			}
+			value, err1 = url.QueryUnescape(value)
+			if err1 != nil {
+				if err == nil {
+					err = err1
```
Comment (shenli) [inline server/region_handler.go]: ditto

File: server/region_handler.go (hunk 11 above)
Comment (jackysp) [reply server/region_handler.go]: OK. I just follow the style of url package. Fixed.

File: server/region_handler.go (hunk 12 above)
Comment (jackysp) [reply server/region_handler.go]: Done.

File: server/region_handler.go (hunk 10 above)
Comment (jackysp) [reply server/region_handler.go]: Add error traces for the whole file.

File: server/region_handler.go (hunk 9 above)
Comment (jackysp) [reply server/region_handler.go]: Done.

**Comment by shenli** [comment]
LGTM

**Pull request merged**